python/ray/air/utils/data_batch_conversion_utils.py

```diff
@@ -0,0 +1,88 @@
+from typing import Type
+
+import numpy as np
+import pandas as pd
+
+from ray.air.data_batch_type import DataBatchType
+from ray.air.constants import TENSOR_COLUMN_NAME
+from ray.util.annotations import DeveloperAPI
+
+try:
+    import pyarrow as pa
+except ImportError:
+    pass
+
+
+@DeveloperAPI
+def convert_batch_type_to_pandas(data: DataBatchType) -> pd.DataFrame:
+    """Convert the provided data to a Pandas DataFrame.
+
+    Args:
+        data: Data of type DataBatchType
+
+    Returns:
+        A pandas Dataframe representation of the input data.
+
+    """
+    from ray.air.utils.tensor_extensions.pandas import TensorArray
+
+    if isinstance(data, pd.DataFrame):
+        return data
+
+    elif isinstance(data, np.ndarray):
+        if data.dtype == object:
+            return pd.DataFrame({TENSOR_COLUMN_NAME: list(data)})
+        else:
+            return pd.DataFrame({TENSOR_COLUMN_NAME: TensorArray(data)})
+
+    elif isinstance(data, dict):
+        tensor_dict = {}
+        for k, v in data.items():
+            # Convert numpy arrays to TensorArray.
+            tensor_dict[k] = TensorArray(v)
+        return pd.DataFrame(tensor_dict)
+
+    elif isinstance(data, pa.Table):
+        return data.to_pandas()
+    else:
+        raise ValueError(
+            f"Received data of type: {type(data)}, but expected it to be one "
+            f"of {DataBatchType}"
+        )
+
+
+@DeveloperAPI
+def convert_pandas_to_batch_type(
+    data: pd.DataFrame, type: Type[DataBatchType]
+) -> DataBatchType:
+    """Convert the provided Pandas dataframe to the provided ``type``.
+
+    Args:
+        data: A Pandas DataFrame
+        type: The specific ``DataBatchType`` to convert data type.
+
+    Returns:
+        The input data represented with the provided type.
+    """
+    if type == pd.DataFrame:
+        return data
+
+    elif type == np.ndarray:
+        # If just a single column, return as a single tensor.
+        if len(data.columns) == 1:
+            return data.iloc[:, 0].to_numpy()
```

Comment:
Should we require the column to be named __value__?

Reply:
My preference is not have this type of distinction, and to always squeeze for single column dataframes.

Reply:
No, just that the former should be returned as `{"x": array([1, 2])}`. We only unpack it as a single numpy array if (1) there is a single column, and (2) that single column name is `__value__`.

Reply:
Hmm but this breaks the API that the output will always be converted back to the same type as the input. If the input type is a numpy array, then in the current API we have to convert the predictions output dataframe to back to a numpy array, not a dict.

The only options are whether to squeeze or unsqueeze as mentioned above.

Reply:
This doesn't seem practical for numpy: what if the model returned two tensors? Then it cannot be joined into a single tensor no matter what.

I think for numpy we have to either return dict or single array based on the model output.

Reply:
Agreed, this is not perfect. Though this will still work- the tensors will be combined as object dtype.

```
>>> from ray.air.util.tensor_extensions.pandas import TensorArray
>>> import pandas as pd
>>> df = pd.DataFrame({"x": TensorArray([1, 2, 3]), "y": TensorArray([[1, 1], [2, 2], [3, 3]])})
>>> df.to_numpy()
array([[1, array([1, 1])],
       [2, array([2, 2])],
       [3, array([3, 3])]], dtype=object)
```

Reply:
The other alternative is to _always_ return a Pandas Dataframe.

But in general I would like to have stronger guarantees on the output type. Either
1. Always return as same type as input
2. Always return pd.Dataframe (or pyarrow.Table for _predict_arrow)

I would prefer not to do something like option 1, but then special case numpy input which may have a dict output. For example, this leads to problems with BatchPredictor if the input batch format is a numpy array, but the output is dict, since Datasets does not support dict batch format right?

Reply:
Why not:

1. Always return as same type as input, but for numpy this can be either (1) single numpy or (2) dict of numpy.

That should satisfy all the use cases. For BatchPredictor, we can convert dict of numpy to pandas explicitly.

Reply:
Ok that sounds fine.

Reply:
Serve wants exactly the same type so we can do batching and splitting. But we also have a last escape hatch to just take a list of object to get around the default splitting behavior.

i.e. the following is supported:
* array inp -> array out
* df inp -> df out
* array inp -> list out
* df inp -> list out

but "array inp -> dict of array out" is definitely not supported atm. (how do you split things back?)

Reply:
here's the code for batching and splitting for np array, you see the error checking we are doing here:
https://github.com/ray-project/ray/blob/master/python/ray/serve/model_wrappers.py#L39-L57